src/linear_algebra/finsupp.lean

```diff
@@ -288,35 +289,67 @@ begin
   exact linear_map.is_linear _
 end
 
+section lsum
+
+variables (S) [semimodule S N] [smul_comm_class R S N]
+
 /-- Lift a family of linear maps `M →ₗ[R] N` indexed by `x : α` to a linear map from `α →₀ M` to
 `N` using `finsupp.sum`. This is an upgraded version of `finsupp.lift_add_hom`.
-We define this as an additive equivalence. For a commutative `R`, this equivalence can be
-upgraded further to a linear equivalence. -/
-def lsum : (α → M →ₗ[R] N) ≃+ ((α →₀ M) →ₗ[R] N) :=
+
+See note [bundled maps over different rings] for why separate `R` and `S` semirings are used.
+-/
+def lsum : (α → M →ₗ[R] N) ≃ₗ[S] ((α →₀ M) →ₗ[R] N) :=
 { to_fun := λ F, {
     to_fun := λ d, d.sum (λ i, F i),
     map_add' := (lift_add_hom (λ x, (F x).to_add_monoid_hom)).map_add,
     map_smul' := λ c f, by simp [sum_smul_index', smul_sum] },
   inv_fun := λ F x, F.comp (lsingle x),
   left_inv := λ F, by { ext x y, simp },
   right_inv := λ F, by { ext x y, simp },
-  map_add' := λ F G, by { ext x y, simp } }
+  map_add' := λ F G, by { ext x y, simp },
+  map_smul' := λ F G, by { ext x y, simp } }
 
-@[simp] lemma coe_lsum (f : α → M →ₗ[R] N) : (lsum f : (α →₀ M) → N) = λ d, d.sum (λ i, f i) := rfl
+@[simp] lemma coe_lsum (f : α → M →ₗ[R] N) : (lsum S f : (α →₀ M) → N) = λ d, d.sum (λ i, f i) :=
+rfl
 
 theorem lsum_apply (f : α → M →ₗ[R] N) (l : α →₀ M) :
-  finsupp.lsum f l = l.sum (λ b, f b) := rfl
+  finsupp.lsum S f l = l.sum (λ b, f b) := rfl
 
 theorem lsum_single (f : α → M →ₗ[R] N) (i : α) (m : M) :
-  finsupp.lsum f (finsupp.single i m) = f i m :=
+  finsupp.lsum S f (finsupp.single i m) = f i m :=
 finsupp.sum_single_index (f i).map_zero
 
-theorem lsum_symm_apply (f : (α →₀ M) →ₗ[R] N) (x : α) : lsum.symm f x = f.comp (lsingle x) := rfl
+theorem lsum_symm_apply (f : (α →₀ M) →ₗ[R] N) (x : α) :
+  (lsum S).symm f x = f.comp (lsingle x) := rfl
+
+end lsum
+
+section
+variables (M) (R) (X : Type*)
+
+/--
+A slight rearrangement from `lsum` gives us
+the bijection underlying the free-forgetful adjunction for R-modules.
+-/
+noncomputable def lift : (X → M) ≃+ ((X →₀ R) →ₗ[R] M) :=
```

Comment:
If you're feeling ambitious this could be a `≃ₗ[S]` too, probably; but you'd need to add `linear_equiv.arrow_congr`, and it's probably out of scope for this PR.

Reply:
It's getting late. I think I'll declare that out of scope for this PR.